### ISSUE TITLE
Mention notifications triggered for System messages for other user with similar name (join channel/team)

### ISSUE THREAD

**Issue by aarowman** [mattermost/mattermost#19031]
Per Mattermost guidelines, GitHub issues are for bug reports: <http://www.mattermost.org/filing-issues/>.

#### Summary
Mention notifications are triggered for System messages for other user with similar name (join channel/team).

#### Version Info
Client Version 5.0.1
~~Server Version 5.39.2~~
Server Version 5.38.2 (Team Edition)

#### Steps to reproduce
1. Set up global mentions (`Other non case-sensitive words, separated by commas`) to include your first name or nickname (e.g. "john" for John Doe)
2. Have someone else with that name (e.g. "John Smith") join or leave a channel/team you're already in
3. Receive mention notification from System alert "John Smith joined the [channel/team]"

#### Expected behavior
Do not receive a mention notification from System alert messages

#### Observed behavior (that appears unintentional) 
Received mention notification from System alert message (and got an email for missed mention if not online)
- this happens multiple times if situation occurs in multiple channels

#### What to fix

We will need to modify the notification code on the server for notifications that checks if a user was mentioned, to ignore keywords and anything else that isn’t your specific username or name when the message is a system message. We should also check that the front-end code is not generating notifications for this either.

**Comment by amyblais** [comment]
@aarowman Can you help confirm your Mattermost server version? You can find it from Main Menu > About Mattermost.

**Comment by aarowman** [comment]
@amyblais Updated issue with better info.
Server: 5.39.2 (Mattermost Team Edition)
Client: 5.0.1

**Comment by amyblais** [comment]
Thank you, I wasn't able to reproduce this just now. Are you able to reproduce this on our [community server here](https://community.mattermost.com/)? (It is on the latest version)

**Comment by aarowman** [comment]
We are actually 5.38.2 (There's a UI bug that displays as 5.39.2).

I'd have to create an account there and have someone else with the same display name join a channel. We're on an enterprise install for my company.

Was this bug fixed in a recent version? I don't see where to check

**Comment by amyblais** [comment]
@aarowman 
1. What are the usernames of the sample users?
2. What are the other preferences set to in Notifications > Words That Trigger Mentions?

**Comment by aarowman** [comment]
@amyblais The description should give that info. I'll elaborate:

Let's say we have user 1: 
* john.c.doe
* display name "John C Doe" 
* and his Words That Trigger Mentions has "John"

Then user 2:
* john.z.smith
* display name "John Z Smith"

User 2 joins a channel that user 1 was already in. User 1 gets a mention notification because the system message says "`John` Z Smith joined the channel"

**Comment by amyblais** [comment]
@lindy65 Would you be able to help test if you can reproduce this?

**Comment by lindy65** [comment]
@amyblais - I created a test environment as per the reporter's usernames and trigger words but could not reproduce the issue

**Comment by aarowman** [comment]
@lindy65 did user 1 have the channel already open? The mention wouldn't trigger a notification if the channel was open. If you move to a new channel, then it may come up (also consider delays).

I just tested it again by adding a new name that triggers notification (case insensitive option) and asking a friend with that name to join a channel. After a minute, it showed the (1) badge for me, even though I had system notifications hidden. (When I opened the channel, there were no comments for over a week, so I knew it wasn't due to another comment triggering it)

**Comment by lindy65** [comment]
Hi @aarowman - no, user 1 did not have the channel in view/open.  Can you join this server [here](https://notif.test.mattermost.cloud/signup_user_complete/?id=yzp334x6k7b7byo1jo7859wrbc) and try to reproduce your issue?  I've tried another few scenarios and still can't reproduce your issue.

Can you tell me whether you have any other words in `Other non case-sensitive words, separated by commas:` (Notifications > Words that trigger mentions)

**Comment by lindy65** [comment]
@aarowman - if you have the word `john` in `Other non case-sensitive words, separated by commas: ` then you will receive notifications whenever the word `john` is mentioned - whether it is for your user or not because it is simply notifying you of the instance of the word `john` - could this be what you're seeing?

![image](https://user-images.githubusercontent.com/13020833/143076929-6f87ef66-542e-41d7-ac97-0a71fc5ba44d.png)

![image](https://user-images.githubusercontent.com/13020833/143077243-e9723c6f-7e19-4688-9613-6dd6d4fd3c1d.png)

**Comment by aarowman** [comment]
Yes! That is exactly what I'm seeing.. I would expect the System notifications for a different person should NOT trigger notifications, even the case insensitive "other" words. That is the exact scenario.

**Comment by aarowman** [comment]
I updated the description to clarify that `Other non case-sensitive words, separated by commas` is the key step to reproduce

**Comment by lindy65** [comment]
Aaah, thanks for clarifying @aarowman.  I don't believe the purpose of `Other non case-sensitive words, separated by commas` is to list your own username or first name...  I believe the feature is working as intended and what you're seeing is not a bug. Hope the below helps?

![image](https://user-images.githubusercontent.com/13020833/143082307-0f2df3d2-8478-47a5-b9d2-8652afbe9caa.png)

**Comment by aarowman** [comment]
Hmm, that's not very clear from the way it's phrased... In my case, I kept missing notifications when someone talked to me because they didn't capitalize my name.

Other people may have a nick name that they want (e.g. user name katherine.a.plankton, wants to be notified for "Kate"). Then if someone else has "Kate" as their display name or first name, then it still triggers)

**Comment by aarowman** [comment]
Either way, I think a system notification still shouldn't trigger this mention. What if someone's name is "Bill Meeting" - then you get an alert for 'meeting' ? The intent behind a system notification is different and I think it shouldn't trigger direct mentions at all.

**Comment by aarowman** [comment]
Is there any situation where we WANT the system notification to do a direct mention to a user?

**Comment by lindy65** [comment]
Thanks @aarowman - @amyblais perhaps a product manager can weigh in here to address the above scenarios/questions?

**Comment by amyblais** [comment]
cc @esethna The observed behaviour is that System Messages trigger a notification when another user with a similar name loins/leaves a channel. We're not sure if this is expected.

**Comment by aarowman** [comment]
I tested this out with system notifications hidden, and it still triggers the alert badge. At a minimum, that should NOT happen. I think all system messages shouldn't do it though, whether hidden or not.
It may be just join. A quick test didn't show the alert for leaving a channel (at least with system notifications hidden).

Don't forget email notifications that get triggered too

**Comment by esethna** [comment]
I agree that system messages for join/leave/added/removed messages should only trigger notifications for the user that's joining/leaving. Let's file a big for that.

Separately let's file a bug for being notified when the system messages are off, that should be an easier fix I imagine. 

cc// @amyblais @lindy65

**Comment by amyblais** [comment]
Created tickets https://mattermost.atlassian.net/browse/MM-40319 & https://mattermost.atlassian.net/browse/MM-40318

**Comment by esethna** [comment]
Thanks, closing to track in the tix

**Comment by aarowman** [comment]
Hello, any updates on this? It has been nearly a year and it looks like the tickets have not had any updates at all. 

In my opinion, we should not close the issue here if it has not been fixed, but I was hoping that the ticket would get prioritized and worked at some point! It is currently `unscheduled`

**Comment by esethna** [comment]
@aarowman this does not seem to be reproducible on master: https://mattermost.atlassian.net/browse/MM-40319?focusedCommentId=136274

**Comment by aarowman** [comment]
@esethna Ok, and that link is specifically for disabled system messages. I haven't tested that lately myself. 
How about when system messages are enabled? That still happens for me on latest version (on prem deployment).

I'm guessing the fix for the second will fix both anyway

**Comment by jgilliam17** [comment]
Thanks @aarowman for your patience while we tried to reproduce the issue.
I can reproduce on latest master, added my findings to https://mattermost.atlassian.net/browse/MM-40318

**Comment by devinbinnie** [comment]
Reopened as HW.

**Comment by costa-neto** [comment]
Is this still an issue? Can I work on it?

**Comment by aarowman** [comment]
It's still happening for me, yes
